static/js/common.js

```diff
@@ -14,27 +14,39 @@ function autofocus(selector) {
 //
 // This is in common.js because we want to use it from the signup page
 // and also from the in-app password change interface.
-function password_quality(password, bar) {
+function password_quality(password, bar, password_field) {
     // We load zxcvbn.js asynchronously, so the variable might not be set.
     if (typeof zxcvbn === 'undefined') {
         return undefined;
     }
 
+    var min_length = 6;
+    var min_quality = 0;
+
+    if (password_field) {
+        min_length = password_field.data('minLength') || min_length;
+        min_quality = password_field.data('minQuality') || min_quality;
+    }
+
     // Consider the password acceptable if it's at least 6 characters.
-    var acceptable = password.length >= 6;
+    var acceptable = password.length >= min_length;
 
-    if (bar !== undefined) {
-        // Compute a quality score in [0,1].
-        var result  = zxcvbn(password);
-        var quality = Math.min(1,Math.log(1 + result.crack_times_seconds.
-                                              offline_slow_hashing_1e4_per_second) / 22);
+    // Compute a quality score in [0,1].
+    var result  = zxcvbn(password);
+    var quality = Math.min(1,Math.log(1 + result.crack_times_seconds.
+                                          offline_slow_hashing_1e4_per_second) / 22);
 
-        // Even if zxcvbn loves your short password, the bar should be filled
-        // at most 1/3 of the way, because we won't accept it.
-        if (!acceptable) {
-            quality = Math.min(quality, 0.33);
-        }
+    // Even if zxcvbn loves your short password, the bar should be filled
+    // at most 1/3 of the way, because we won't accept it.
+    if (!acceptable) {
+        quality = Math.min(quality, 0.33);
```

Comment:
This actually fills the bar 39.7% of the way due to the 10% hack later in the function.